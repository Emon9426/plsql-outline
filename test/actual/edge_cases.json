{
  "nodes": [
    {
      "type": "PROCEDURE",
      "name": "edge_case_procedure",
      "declarationLine": 2,
      "beginLine": 46,
      "exceptionLine": 65,
      "endLine": 68,
      "level": 1,
      "children": [
        {
          "type": "PROCEDURE",
          "name": "simple_end",
          "declarationLine": 6,
          "beginLine": 7,
          "exceptionLine": null,
          "endLine": 9,
          "level": 2,
          "children": []
        },
        {
          "type": "FUNCTION",
          "name": "func_with_named_end",
          "declarationLine": 12,
          "beginLine": 13,
          "exceptionLine": null,
          "endLine": 15,
          "level": 2,
          "children": []
        },
        {
          "type": "PROCEDURE",
          "name": "mixed_end",
          "declarationLine": 18,
          "beginLine": 28,
          "exceptionLine": null,
          "endLine": 35,
          "level": 2,
          "children": [
            {
              "type": "FUNCTION",
              "name": "inner_func",
              "declarationLine": 19,
              "beginLine": 20,
              "exceptionLine": null,
              "endLine": 22,
              "level": 3,
              "children": []
            },
            {
              "type": "PROCEDURE",
              "name": "inner_proc",
              "declarationLine": 24,
              "beginLine": 25,
              "exceptionLine": null,
              "endLine": 27,
              "level": 3,
              "children": []
            }
          ]
        }
      ]
    }
  ],
  "metadata": {
    "sourceFile": "edge_cases.prc",
    "totalLines": 45,
    "maxNestingDepth": 3
  }
}