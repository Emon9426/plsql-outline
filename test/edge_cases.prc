/* 边界情况测试 */
CREATE OR REPLACE PROCEDURE edge_case_procedure AS
  /* 注释中的关键字：BEGIN END FUNCTION PROCEDURE DECLARE */
  
  -- 情况1：只有END
  PROCEDURE simple_end IS
  BEGIN
    NULL;
  END; -- 只有END
  
  -- 情况2：END后跟函数名
  FUNCTION func_with_named_end RETURN NUMBER IS
  BEGIN
    RETURN 1;
  END func_with_named_end; -- END后跟函数名
  
  -- 情况3：多层嵌套与混合END
  PROCEDURE mixed_end IS
    FUNCTION inner_func RETURN NUMBER IS
    BEGIN
      RETURN 2;
    END inner_func; -- END后跟函数名
    
    PROCEDURE inner_proc IS
    BEGIN
      NULL;
    END; -- 只有END
  BEGIN
    DECLARE
      v_temp NUMBER := 0;
    BEGIN
      v_temp := inner_func();
      DBMS_OUTPUT.PUT_LINE('值: ' || v_temp);
    END; -- 只有END
  END mixed_end; -- END后跟过程名
  
  -- 情况4：注释中的嵌套结构
  /*
    FUNCTION commented_function IS
    BEGIN
      -- 这只是一个注释
      NULL;
    END commented_function;
  */
  
BEGIN
  -- 主执行块
  DECLARE
    v_result NUMBER;
  BEGIN
    v_result := func_with_named_end();
    DBMS_OUTPUT.PUT_LINE('函数结果: ' || v_result);
    mixed_end();
    
    -- 异常块测试
    BEGIN
      RAISE NO_DATA_FOUND;
    EXCEPTION
      WHEN NO_DATA_FOUND THEN
        DBMS_OUTPUT.PUT_LINE('捕获到NO_DATA_FOUND异常');
    END; -- 只有END
  END; -- 只有END
  
  /* 结束注释 */
EXCEPTION
  WHEN OTHERS THEN
    DBMS_OUTPUT.PUT_LINE('主异常块: ' || SQLERRM);
END edge_case_procedure; -- END后跟过程名
/
