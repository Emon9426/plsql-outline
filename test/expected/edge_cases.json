{
  "name": "edge_case_procedure",
  "type": "PROCEDURE",
  "children": [
    {
      "name": "simple_end",
      "type": "PROCEDURE",
      "children": [
        {
          "name": "BEGIN",
          "type": "BLOCK"
        }
      ]
    },
    {
      "name": "func_with_named_end",
      "type": "FUNCTION",
      "children": [
        {
          "name": "BEGIN",
          "type": "BLOCK"
        }
      ]
    },
    {
      "name": "mixed_end",
      "type": "PROCEDURE",
      "children": [
        {
          "name": "inner_func",
          "type": "FUNCTION",
          "children": [
            {
              "name": "BEGIN",
              "type": "BLOCK"
            }
          ]
        },
        {
          "name": "inner_proc",
          "type": "PROCEDURE",
          "children": [
            {
              "name": "BEGIN",
              "type": "BLOCK"
            }
          ]
        },
        {
          "name": "BEGIN",
          "type": "BLOCK",
          "children": [
            {
              "name": "DECLARE",
              "type": "DECLARE"
            },
            {
              "name": "BEGIN",
              "type": "BLOCK"
            }
          ]
        }
      ]
    },
    {
      "name": "BEGIN",
      "type": "BLOCK",
      "children": [
        {
          "name": "DECLARE",
          "type": "DECLARE"
        },
        {
          "name": "BEGIN",
          "type": "BLOCK",
          "children": [
            {
              "name": "BEGIN (异常测试块)",
              "type": "BLOCK"
            },
            {
              "name": "EXCEPTION",
              "type": "EXCEPTION"
            }
          ]
        }
      ]
    },
    {
      "name": "EXCEPTION",
      "type": "EXCEPTION"
    }
  ]
}
